packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -236,7 +352,85 @@ contract MetaTransactionWallet is
   }
 
   /**
-   * @notice Executes a transaction on behalf of the signer.`
+  * relayer? TODO determine if we can predict this request (question for komenci folks)
+  * gatewayFee / gateWayFeeRecipient? (TODO: ask Contracts if better to omit these or add them for forward compatibility) 
+  */
+
+  /**
+   * @notice Executes a refundable meta-transaction on behalf of the signer.
+   * @param destination The address to which the meta-transaction is to be sent.
+   * @param value The CELO value to be sent with the meta-transaction.
+   * @param data The data to be sent with the meta-transaction.
+   * @param maxGasPrice The maximum gas price the user is willing to pay.
+   * @param gasLimit The gas limit for entire relayed transaction.
+   * @param metaGasLimit The gas limit for just the meta-transaction.
+   * @param v The recovery id of the ECDSA signature of the meta-transaction.
+   * @param r Output value r of the ECDSA signature.
+   * @param s Output value s of the ECDSA signature.
+   * @return The return value of the meta-transaction execution.
+   */
+  function executeMetaTransactionWithRefund(
+    address destination,
+    uint256 value,
+    bytes calldata data,
+    uint256 maxGasPrice,
+    uint256 gasLimit,
+    uint256 metaGasLimit,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) external returns (bytes memory) {
+    require(gasLimit == gasLeft(), "gasLimit different than limit authorized by signer"); // To ensure Komenci actually set the correct gas limit TODO ask Contracts about this
+    require(tx.gasprice <= maxGasPrice, "gasprice exceeds limit authorized by signer");
+    require(address(this).balance >= gasLimit.mul(tx.gasprice).add(value), "insufficient balance");
+    require(msg.sender == tx.origin, "relayer must be EOA");
+    require(gasLimit <= block.gaslimit, "Impossible gas limit");
+    require(metaGasLimit < gasLimit, "metaGasLimit must be less than gasLimit");
+    if (data.length > 0) require(Address.isContract(destination), "Invalid contract address");
```

Comment:
hmm i wonder if you should revert on this. if they are calling a non-contract address with data, it almost certainly means they have submitted a malformed meta-transaction, but I wonder whether it would be better to remain un-opinionated about the contents of the meta-transaction - let users execute a messed up tx if they sign it & pay for it ¯\_(ツ)_/¯